Added initial bats tests and github action set up

gaol opened the pull request:
This PR tries to add initial bats tests for Hera and set up github action CI.

* `tests/tests-common.sh` - based on the one in harmonia
* `tests/tests-suite.sh` - based on the one in harmonia

spyrkob commented on tests/run.sh-tests.bats:
```diff
@@ -0,0 +1,43 @@
+
+teardown() {
+  deleteIfExist "${WORKSPACE}"
+  unset JBOSS_FOLDER
```
where is that set?

gaol replied:
Right, this is copied from harmonia, it should be amended for hera. thanks :)

spyrkob commented on tests/tests-suite.sh:
```diff
@@ -0,0 +1,37 @@
+do
+  echo "<h1>${script_file}</h1>" >> "${shellcheck_report}"
+  shellcheck --format=quiet "${script_file}"
+  if [ "${?}" -ne 0 ] ; then
```
If we find spellcheck violations should we return non-zero code from this script to fail github action?

gaol replied:
You are right, the same should for the bats tests running. :+1:

spyrkob commented on tests/tests-common.sh:
```diff
@@ -0,0 +1,79 @@
+  readonly DUMMY_COMMAND_DIR=${DUMMY_COMMAND_DIR:-$(mktemp -d)}
+  export DUMMY_COMMAND_DIR
+  trap 'deleteIfExist ${DUMMY_COMMAND_DIR}' EXIT
+  export PATH=${DUMMY_COMMAND_DIR}:${PATH}
```
think that should be "${DUMMY_COMMAND_DIR}":"${PATH}"

gaol replied:
Done

spyrkob commented on tests/tests-common.sh:
```diff
@@ -0,0 +1,79 @@
+
+setupDummySSH() {
+  export DUMMY_SSH=${DUMMY_SSH:-"$(pwd)/tests/"}
+  export PATH=${DUMMY_SSH}:${PATH}
```
same as above

spyrkob reviewed (commented):
couple small questions, otherwise LGTM... at least for my bash-fu ;)

gaol commented:
Feedback on the review:
* Updated `tests-suite.sh ` accordingly to fail the github action when either bats test fails or ShellCheck violations are found.
* Updated `job.sh` to fix ShellCheck violation of [SC2145](https://github.com/koalaman/shellcheck/wiki/SC2145)
* Added a `.shellcheckrc` file to exclude some ShellCheck rules to pass the CI for the moment. I am not sure how to solve the violations, as currently all work well, I think this should be OK to ignore currently.

spyrkob commented on job.sh:
```diff
@@ -46,4 +46,4 @@ run_ssh "podman exec \
         -e VERSION="${VERSION}" \
         -e COMPONENT_UPGRADE_LOGGER="${COMPONENT_UPGRADE_LOGGER}" \
-        -ti ${CONTAINER_NAME} '${BUILD_SCRIPT}' ${@}"
+        -ti ${CONTAINER_NAME} '${BUILD_SCRIPT}' $*"
```
I think it should be OK, but @rpelisse could you confirm this?

rpelisse replied:
@spyrkob @gaol I'm not sure about this change, I've never seen $* before... What is the reason for this change? What does $* that differs from ${@} ? In any case, I think it should be ${*}.

spyrkob commented:
Thanks @gaol! 

I think running shellcheck with `-x` to solve SC1091 (ie allow reading linked files)
The rest of violations seem to be valid points (AFAICT), we should address those, but it can be separate PR.

rpelisse commented on tests/tests-suite.sh:
```diff
@@ -0,0 +1,54 @@
+  done
+  exit 1
+fi
+echo 'Done - PASSED'
```
@gaol There seems to be quite a lot of code duplication from Harmonia's testsuite... I wonder if it would no be cleaner to have Jenkins checkout Harmonia inside the Hera folder and reuse the stuff... WDTK?

rpelisse commented:
@gaol where are we on this PR? It seems to have gone stale, can you rebase and update?

gaol commented:
Ah, it must had been lost in my list :grin:  will update it soon
